benches/nexmark/nexmark_dbsp_source.rs

```diff
@@ -0,0 +1,120 @@
+//! DBSP Source operator that reads from a Nexmark Generator.
+
+use crate::generator::{NexmarkGenerator, NextEvent};
+use dbsp::{
+    algebra::{ZRingValue, ZSet},
+    circuit::{
+        operator_traits::{Data, Operator, SourceOperator},
+        Scope,
+    },
+};
+use rand::Rng;
+use std::{borrow::Cow, marker::PhantomData};
+
+pub struct NexmarkDBSPSource<R: Rng, W, C> {
+    /// The generator for events emitted by this source.
+    // TODO(absoludity): Longer-term, it'd be great to use a client (such as a gRPC client) here to
+    // completely separate the generator and allow the generator to be used with other languages
+    // etc.
+    generator: NexmarkGenerator<R>,
+    batch_size: usize,
+    _t: PhantomData<(C, W)>,
+}
+
+impl<R, W, C> NexmarkDBSPSource<R, W, C>
+where
+    R: Rng,
+{
+    pub fn from_generator(generator: NexmarkGenerator<R>, batch_size: usize) -> Self {
+        NexmarkDBSPSource {
+            generator,
+            batch_size,
+            _t: PhantomData,
+        }
+    }
+}
+
+impl<R, W, C> Operator for NexmarkDBSPSource<R, W, C>
+where
+    C: Data,
+    R: Rng + 'static,
+    W: 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("NexmarkDBSPSource")
+    }
+
+    // Do clock_start and clock_end make sense for an infinite source? Or should
+    // the Nexmark generator use the max_events and stop generating events after
+    // that.
+    /// That would also mean then that the fixed point for this data source
+    /// would be at max_events?  Currently the generator will keep spitting out
+    /// events, but maybe `next_event` could be updated to return a
+    /// `Result<Option<NextEvent>>` and return Ok(None) after max events?
+    fn fixedpoint(&self, _scope: Scope) -> bool {
+        false
+    }
+}
+
+impl<R, W, C> SourceOperator<C> for NexmarkDBSPSource<R, W, C>
+where
+    R: Rng + 'static,
+    W: ZRingValue + 'static,
+    C: Data + ZSet<Key = NextEvent, R = W>,
+{
+    fn eval(&mut self) -> C {
+        // Q: Why is the time argument expecting ()? Same for the key tuple's second el?
```

Comment:
// Q: Why is the time argument expecting ()? Same for the key tuple's second el?

Notionally, all DBSP batches are collections of 4-tuple: `(key, value, time, weight)`. The `Batch::from_tuples(time, tuples)` method creates a batch where all tuples have the same `time` value, so the `tuples` vector only needs to specify `key`, `value`, and `weight`.  For Z-sets, the `value` type is `()`.

As for the time type being `()`, it's a form of optimization that may or may not survive future design iterations :) Technically, each circuit has a clock that increments sequentially from zero to infinity.  We could always model such a clock as `usize`.  A nested circuit has its own local clock, so its logical time can be modeled as a `(usize, usize)` tuple.  Every tuple in a batch is labeled by the logical time when it war produced: tuples produced during the first clock cycle have timestamp 0, tuples produced during the second cycle have timestamp 1, etc.

It turns out though that we don't need to track the time of each tuple precisely.  We need to store just enough timing information to distinguish between tuples currently in flight inside the circuit.  In our current "synchronous" model where we process each input to completion before ingesting more data, we don't need to track the timing at all in the root circuit, hence the `()` time.

There are some serious questions about the ability of this architecture to scale out, so the design may change in the future.